platforms/chibios/drivers/ws2812_pio.c

```diff
@@ -0,0 +1,138 @@
+// Copyright 2022 Stefan Kerkmann
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "quantum.h"
+#include "ws2812.h"
+#include "hardware/pio.h"
+#include "hardware/clocks.h"
+
+#if !defined(MCU_RP)
+#    error PIO Driver is only available for Raspberry Pi 2040 MCUs!
+#endif
+
+#if defined(WS2812_PIO_USE_PIO1)
+static const PIO pio = pio1;
+#else
+static const PIO pio = pio0;
+#endif
+
+static int state_machine = -1;
+
+#define WS2812_WRAP_TARGET 0
+#define WS2812_WRAP 3
+
+#define WS2812_T1 2
+#define WS2812_T2 5
+#define WS2812_T3 3
+
+#if defined(WS2812_EXTERNAL_PULLUP)
+
+#    pragma message "The GPIOs of the RP2040 are NOT 5V tolerant! Make sure to NOT apply any voltage over 3.3V to the RGB data pin."
```

Comment:
Is the WS2812 back-driving a signal to the RP2040 or is it a strictly one-way communication?
If only the RP2040 is driving the pin, and the WS2812 is only "reading", is there any danger?

Reply:
No the WS2812 are data in only, but they operate at 5V supply voltage and logic levels. So to remedy the 3.3V logic level of most ARM chips, a cheap way to do 3.3V to 5V logic level conversion is to use an open-drain bus configuration. In this configuration an external pull-up resistors determinants the high level of the bus (5V), which the output pin of the rp2040 still is exposed to.